crates/flowistry/benches/main.rs

```diff
@@ -0,0 +1,181 @@
+#![feature(rustc_private)]
+
+extern crate rustc_borrowck;
+extern crate rustc_driver;
+extern crate rustc_hir;
+extern crate rustc_interface;
+extern crate rustc_middle;
+
+use std::process::Command;
+
+use criterion::{criterion_group, criterion_main, Criterion};
+use flowistry::{
+  infoflow::Direction,
+  mir::{borrowck_facts, utils::PlaceExt},
+};
+use glob::glob;
+use rustc_borrowck::BodyWithBorrowckFacts;
+use rustc_hir::{BodyId, ItemKind};
+use rustc_middle::{
+  mir::{Location, Place},
+  ty::TyCtxt,
+};
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum AnalysisType {
+  FlowOnly,
+  FlowAndDeps,
+}
+
+fn analysis<'tcx>(
+  tcx: TyCtxt<'tcx>,
+  body_id: BodyId,
+  body_with_facts: &BodyWithBorrowckFacts<'tcx>,
+  ty: AnalysisType,
+) {
+  let results = flowistry::infoflow::compute_flow(tcx, body_id, body_with_facts);
+
+  if ty == AnalysisType::FlowAndDeps {
+    let mut targets = vec![];
+
+    for local in body_with_facts.body.local_decls.indices() {
+      let arg = Place::make(local, &[], tcx);
+      targets.push(vec![(arg, Location::START)]);
+    }
+
+    flowistry::infoflow::compute_dependencies(&results, targets, Direction::Forward);
+  }
+}
+
+struct UnsafeBenchGroup(
+  criterion::BenchmarkGroup<'static, criterion::measurement::WallTime>,
+);
+
+unsafe impl Send for UnsafeBenchGroup {}
+unsafe impl Sync for UnsafeBenchGroup {}
+
+struct Callbacks {
+  ty: AnalysisType,
+  group: UnsafeBenchGroup,
+}
+impl rustc_driver::Callbacks for Callbacks {
+  fn config(&mut self, config: &mut rustc_interface::Config) {
+    config.override_queries = Some(borrowck_facts::override_queries);
+  }
+
+  fn after_parsing<'tcx>(
+    &mut self,
+    _compiler: &rustc_interface::interface::Compiler,
+    queries: &'tcx rustc_interface::Queries<'tcx>,
+  ) -> rustc_driver::Compilation {
+    queries.global_ctxt().unwrap().take().enter(|tcx| {
+      let hir = tcx.hir();
+      let body_id = hir
+        .items()
+        .filter_map(|item| match item.kind {
+          ItemKind::Fn(_, _, body) => Some(body),
+          _ => None,
+        })
+        .next()
+        .unwrap();
+
+      let def_id = hir.body_owner_def_id(body_id);
+      let body_with_facts = borrowck_facts::get_body_with_borrowck_facts(tcx, def_id);
+
+      let bench_id = match self.ty {
+        AnalysisType::FlowOnly => "Flow",
+        AnalysisType::FlowAndDeps => "Flow + Deps",
+      };
+
+      self.group.0.bench_function(bench_id, |b| {
+        b.iter(|| analysis(tcx, body_id, body_with_facts, self.ty))
+      });
+    });
+    rustc_driver::Compilation::Stop
+  }
+}
+
+fn criterion_benchmark(c: &mut Criterion) {
+  let tests = vec![
```

Comment:
Somewhere in this set of files should be instructions for things like -- if you make a new benchmark, how do you add it? And presumably one step is putting a tuple in this vector.

Also, stylistically if this is always just string constants, I would make it a global `static TESTS: ...`